hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/BasicFormatterImpl.java

```diff
@@ -390,6 +201,261 @@ private void newline() {
+							p.inFunction--;
+						}
+						else {
+							if ( !p.afterByOrSetOrFromOrSelect ) {
+								p.indent--;
+								p.newline();
+							}
+						}
+						p.out();
+						p.beginLine = false;
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_BEGIN_NEW_CLAUSE {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( BEGIN_CLAUSES.contains( p.lcToken ) ) {
+						if ( !p.afterBeginBeforeEnd ) {
+							if ( p.afterOn ) {
+								p.indent--;
+								p.afterOn = false;
+							}
+							p.indent--;
+							p.newline();
+						}
+						p.out();
+						p.beginLine = false;
+						p.afterBeginBeforeEnd = true;
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_END_NEW_CLAUSE {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( END_CLAUSES.contains( p.lcToken ) ) {
+						if ( !p.afterBeginBeforeEnd ) {
+							p.indent--;
+							if ( p.afterOn ) {
+								p.indent--;
+								p.afterOn = false;
+							}
+							p.newline();
+						}
+						p.out();
+						if ( !"union".equals( p.lcToken ) ) {
+							p.indent++;
+						}
+						p.newline();
+						p.afterBeginBeforeEnd = false;
+						p.afterByOrSetOrFromOrSelect = "by".equals( p.lcToken )
+								|| "set".equals( p.lcToken )
+								|| "from".equals( p.lcToken );
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_SELECT {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( "select".equals( p.lcToken ) ) {
+						p.out();
+						p.indent++;
+						p.newline();
+						p.parenCounts.addLast( p.parensSinceSelect );
+						p.afterByOrFromOrSelects.addLast( p.afterByOrSetOrFromOrSelect );
+						p.parensSinceSelect = 0;
+						p.afterByOrSetOrFromOrSelect = true;
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_UPDATE_INSERT_DELETE {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( DML.contains( p.lcToken ) ) {
+						p.out();
+						p.indent++;
+						p.beginLine = false;
+						if ( "update".equals( p.lcToken ) ) {
+							p.newline();
+						}
+						if ( "insert".equals( p.lcToken ) ) {
+							p.afterInsert = true;
+						}
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_VALUES {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( "values".equals( p.lcToken ) ) {
+						p.indent--;
+						p.newline();
+						p.out();
+						p.indent++;
+						p.newline();
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_ON {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( "on".equals( p.lcToken ) ) {
+						p.indent++;
+						p.afterOn = true;
+						p.newline();
+						p.out();
+						p.beginLine = false;
+						return true;
+					}
+					return false;
+				}
+			},
+
+			PROCESS_AFTER_BETWEEN_AND {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( p.afterBetween && p.lcToken.equals( "and" ) ) {
```

Comment:
Using `"and".equals( p.lcToken )` would be more consistent and robust, IMHO.

Reply:
Yeah you're right, I'll fix it later.